lms/util/h.py

```diff
@@ -0,0 +1,76 @@
+# -*- coding: utf-8 -*-
+
+"""Utility functions for working with the h API."""
+
+from __future__ import unicode_literals
+
+import random
+import string
+
+
+USERNAME_ALLOWED_CHARS = string.ascii_letters + string.digits + "._"
+"""The characters that are allowed in h usernames."""
+
+
+DISPLAY_NAME_MAX_LENGTH = 30
+"""The maximum length of an h display name."""
+
+
+USERNAME_MAX_LENGTH = 30
+"""The maximum length of an h username."""
+
+
+def generate_display_name(request_params):
+    full_name = (request_params.get("lis_person_name_full") or "").strip()
+    given_name = (request_params.get("lis_person_name_given") or "").strip()
+    family_name = (request_params.get("lis_person_name_family") or "").strip()
+
+    if full_name:
+        display_name = full_name
+    else:
+        display_name = " ".join((given_name, family_name))
+
+    display_name = display_name.strip()
+
+    display_name = display_name or "Anonymous"
+
+    if len(display_name) > DISPLAY_NAME_MAX_LENGTH:
+        display_name = display_name[:DISPLAY_NAME_MAX_LENGTH - 1].rstrip() + "…"
+
+    return display_name
+
+
+def generate_username(request_params):
+    """Return an h username generated from the given LTI launch request params."""
+    full_name = request_params.get("lis_person_name_full") or ""
+    user_id = request_params.get("user_id") or ""
+
+    def n_valid_chars_from(s, n):  # pylint: disable=invalid-name
```

Comment:
I like these here as inner functions

Reply:
Curious why pylint doesn't like this naming? SIngle-letter vars?

Reply:
Pylint doesn't like single-letter variable names. Specifically, pylint's `variable-rgx` setting is a regex that variable names must match in order to be valid, and by default it's `[a-z_][a-z0-9_]{2,30}$` (there are similar `function-rgx`, `method-rgx`, `argument-rgx`, etc etc settings, we customize some of them.) We have a [whitelist of variable names we tell it to allow](https://github.com/hypothesis/lms/blob/64a188cded9920350f719a522bf19e44fe89a88d/.prospector.base.yaml#L37-L54) but I didn't want to add `s` and `n` to that

Reply:
Docs: <http://pylint.pycqa.org/en/latest/user_guide/options.html?highlight=variable-rgx#cmdoption-variable-rgx> (although we would put these settings in our `.prospector.*` config files, not use them as command line args)